### PR TITLE
Add validation of Print format strings

### LINKED ISSUE #2734: Fuzzer issue around Print arguments
### Description of the bug:

Print uses llvm::formatv, which can crash on invalid inputs.

Note, one way of fixing this might be to write custom input string validation that limits content, producing a compilation error instead of a crash.

### What did you do, or what's a simple way to reproduce the bug?

```
package EmptyIdentifier impl;

fn Main() -> i32 {
  Print("{1 1}");
  return 0;
}
```

### What did you expect to happen?

Should error on the format string

### What actually happened?

Crashes

### Any other information, logs, or outputs that you want to share?

Case 269349669 for those with access to the current fuzzer reporting.

### PR DESCRIPTION
**Pull request by jonmeow**
Also adds some validation of various direct print intrinsic calls.

Fixes #2734

### CODE REVIEW

File: explorer/testdata/print/fail_index_bounds1.carbon [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+//
+// AUTOUPDATE
+// RUN: %{not} %{explorer-run}
+// RUN: %{not} %{explorer-run-trace}
+
+package ExplorerTest impl;
+
+fn Main() -> i32 {
+  // CHECK:STDERR: RUNTIME ERROR: {{.*}}/explorer/testdata/print/fail_index_bounds1.carbon:[[@LINE+1]]: Invalid format string for Print with 1 arg(s) : `Print: {0} {1}`
+  Print("Print: {0} {1}", 1);
```
Comment (geoffromer) [inline explorer/testdata/print/fail_index_bounds1.carbon:13]:
```suggestion
  Print("Print: {1}", 1);
```

I initially misread this as being about a mismatch in the _number_ of placeholders vs. arguments, rather than the placeholder index itself being out of range.

File: explorer/testdata/print/fail_intrinsic_int_format.carbon [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+// RUN: %{not} %{explorer-run}
+// RUN: %{not} %{explorer-run-trace}
+
+package ExplorerTest impl;
+
+fn Main() -> i32 {
+  // CHECK:STDERR: COMPILATION ERROR: {{.*}}/explorer/testdata/print/fail_intrinsic_int_format.carbon:[[@LINE+3]]: type error in Print argument 0
+  // CHECK:STDERR: expected: String
+  // CHECK:STDERR: actual: i32
+  __intrinsic_print(0);
```
Comment (geoffromer) [inline explorer/testdata/print/fail_intrinsic_int_format.carbon:15]: The whole point of intrinsics is that they're not supposed to occur in user code; they're just implementation details of the prelude. That being the case, should we test them at all, or just test the prelude wrappers?

File: explorer/testdata/print/fail_index_bounds1.carbon (hunk 1 above)
Comment (jonmeow) [reply explorer/testdata/print/fail_index_bounds1.carbon:13]: Is it okay to keep this? I'd been considering testing for unused indices, but your suggested change would be a barrier to doing that.

File: explorer/testdata/print/fail_intrinsic_int_format.carbon (hunk 2 above)
Comment (jonmeow) [reply explorer/testdata/print/fail_intrinsic_int_format.carbon:15]: Unless the system _prevents_ their use, then it is valid code, and could come up from fuzz testing. As a consequence, they should generally be tested to help ensure there aren't bugs.

File: explorer/testdata/print/fail_index_bounds1.carbon (hunk 1 above)
Comment (geoffromer) [reply explorer/testdata/print/fail_index_bounds1.carbon:13]: Hmm, then how about this instead?

```suggestion
  Print("Print: {0} {100}", 1);
```

File: explorer/interpreter/interpreter.cpp [hunk 3]
```diff
@@ -1087,6 +1087,56 @@ auto Interpreter::CallFunction(const CallExpression& call,
...
+            }
+            if (index >= num_args) {
+              // Out-of-bounds.
+              return false;
+            }
+            // Move past the `}`.
+            ++cursor;
+        }
+        break;
+      default:
```
Comment (maan2003) [inline explorer/interpreter/interpreter.cpp:1141]: `Print("}")` passes but shouldn't. Because `}` needs be to escaped as `}}`

Comment (maan2003) [reply explorer/interpreter/interpreter.cpp:1141]: on reading the docs more carefully, "}" is allowed by `llvm::format_v`. But, it still feels weird.

File: explorer/interpreter/interpreter.cpp [hunk 4]
```diff
@@ -1087,6 +1087,56 @@ auto Interpreter::CallFunction(const CallExpression& call,
...
+            // Reject `{}`.
+            return false;
+          default:
+            int index = 0;
+            while (*cursor != '}') {
+              if (*cursor < '0' || *cursor > '9') {
+                // Non-numeric text, including `\0`.
+                return false;
+              }
+              index = (10 * index) + (*cursor - '0');
```
Comment (maan2003) [inline explorer/interpreter/interpreter.cpp:1128]: doesn't handle overflow.

```
Print("{2147483648}", 1);
```
weirdly prints `2147483648`

File: explorer/interpreter/interpreter.cpp (hunk 3 above)
Comment (jonmeow) [reply explorer/interpreter/interpreter.cpp:1141]: Thanks, but this is not actionable.

File: explorer/testdata/print/fail_index_bounds1.carbon (hunk 1 above)
Comment (jonmeow) [reply explorer/testdata/print/fail_index_bounds1.carbon:13]: While that test would fail, it removes the test of the edge case. [i.e., < vs <=]

Comment (geoffromer) [reply explorer/testdata/print/fail_index_bounds1.carbon:13]: OK, then I'm out of ideas, so I guess we can leave it as it is.

**Review by geoffromer — COMMENTED** [review]
LGTM once the overflow concern is addressed.

File: explorer/testdata/print/fail_index_bounds1.carbon (hunk 1 above)
Comment (jonmeow) [reply explorer/testdata/print/fail_index_bounds1.carbon:13]: I've changed the error messages to be more specific.

File: explorer/interpreter/interpreter.cpp (hunk 4 above)
Comment (jonmeow) [reply explorer/interpreter/interpreter.cpp:1128]: Blocked and added a test.

**Comment by jonmeow** [comment]
Note, this PR is still blocked on review/approval. I think all comments have been addressed.

**Review by geoffromer — APPROVED** [review]
